heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManagerServer.java

```diff
@@ -0,0 +1,298 @@
+    } else {
+      LOG.severe("Unknown kind of request: " + request.getClass().getName());
+    }
+  }
+
+  protected void handleCleanStatefulCheckpointRequest(
+      REQID rid,
+      SocketChannel channel,
+      CheckpointManager.CleanStatefulCheckpointRequest request
+  ) {
+    LOG.info("Got a clean request from " + request.toString() + " host:port "
+        + channel.socket().getRemoteSocketAddress());
+
+    boolean deleteAll = request.hasCleanAllCheckpoints() && request.getCleanAllCheckpoints();
+    Common.StatusCode statusCode = Common.StatusCode.OK;
+
+    try {
+      checkpointsBackend.dispose(topologyName,
+                                 request.getOldestCheckpointPreserved(), deleteAll);
+      LOG.info("Dispose checkpoint successful");
+    } catch (StatefulStorageException e) {
+      LOG.info("Dispose checkpoint failed with error " + e);
+      statusCode = Common.StatusCode.NOTOK;
+    }
+
+    CheckpointManager.CleanStatefulCheckpointResponse.Builder responseBuilder =
+        CheckpointManager.CleanStatefulCheckpointResponse.newBuilder();
+    responseBuilder.setStatus(Common.Status.newBuilder().setStatus(statusCode));
+
+    sendResponse(rid, channel, responseBuilder.build());
+  }
+
+  protected void handleTMasterRegisterRequest(
+      REQID rid,
+      SocketChannel channel,
+      CheckpointManager.RegisterTMasterRequest request
+  ) {
+    LOG.info("Got a register request from TMaster host:port "
+        + channel.socket().getRemoteSocketAddress());
+
+    CheckpointManager.RegisterTMasterResponse.Builder responseBuilder =
+        CheckpointManager.RegisterTMasterResponse.newBuilder();
+
+    if (!request.getTopologyName().equals(topologyName)) {
+      LOG.severe("The register message was from a different topology: "
+          + request.getTopologyName());
+      responseBuilder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.NOTOK));
+    } else if (!request.getTopologyId().equals(topologyId)) {
+      LOG.severe("The register message was from a different topology id: "
+          + request.getTopologyName());
+      responseBuilder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.NOTOK));
+    } else if (connection != null) {
+      // TODO(mfu): Should we do this?
+      LOG.warning("We already have an active connection from the tmaster "
+          + "Closing existing connection...");
+
+      try {
+        connection.close();
+      } catch (IOException e) {
+        LOG.warning("Failed to close connection from: "
+                  + connection.socket().getRemoteSocketAddress());
+      }
+
+      connection = null;
+      responseBuilder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.NOTOK));
+    } else {
+      connection = channel;
+      responseBuilder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.OK));
+    }
+
+    sendResponse(rid, channel, responseBuilder.build());
+  }
+
+  protected void handleStMgrRegisterRequest(
```

Comment:
How is the implementation of this method different from handleTMasterRegisterRequest? They appear to be duplicates.